tempesta_fw/ss_skb.c

```diff
@@ -299,36 +316,43 @@ __split_linear_data(struct sk_buff *skb, struct sk_buff *pskb,
 		/* Not found. Return invalid address, and try next skb. */
 		return (void *)1;
 	}
+
 	/*
-	 * Not enough room in the linear part - put data in a page fragment.
+	 * Data is inserted or deleted in the middle of the linear part,
+	 * or there's insufficient room in the linear part of an SKB to
+	 * insert @len bytes.
+	 *
+	 * Don't bother with skb tail room: if the linear part is large,
+	 * then it's likely that we'll do some smaller data insertions
+	 * later and go by the quick path above. Otherwise, the tail size
+	 * is also small.
 	 *
-	 * Don't bother with skb tail: if the linear part is large, then
-	 * it's likely that we'll do some smaller data insertions and go
-	 * by the quick path above, otherwise the tail size is also small.
+	 * The inserted data is placed in a fragment. The tail part is
+	 * moved to yet another fragment. The linear part is trimmed to
+	 * exclude the deleted data and the tail part.
 	 *
-	 * Do all allocations before moving the fragments to avoid
-	 * complex rollback.
+	 * Do all allocations before moving the fragments to avoid complex
+	 * rollback.
 	 */
-
 	if (alloc) {
 		if (__new_pgfrag(skb, pskb, len, 0, alloc + !!tail_len))
 			return NULL;
 	} else {
-		/*
-		 * Probably we should delete the whole second part
-		 * of the linear data or move it to page fragment.
-		 */
-		tail_len += len;
-		BUG_ON(tail_len < 0);
-		if (tail_len && __extend_pgfrags(skb, pskb, 0, 1))
+		if (__extend_pgfrags(skb, pskb, 0, 1))
```

Comment:
Here is the case where we know for sure that we need to insert a new fragment. There's a case where a whole tail is removed from the linear part. That case is processed before this one, not shown in this patch.

Also, `tail_len` can never be zero. If it was that would mean that we've got `pspt` pointing at the end of data in the linear area, because `tail_len` is calculated as `tail_len = (char *)skb_tail_pointer(skb) - pspt;`